### PR TITLE
AlibabaCloud: adding alibaba cloud platform

### PR DESCRIPTION
**Pull request by kwoodson**
The initial attempt at creating the necessary bits for Alibaba.

### CODE REVIEW

**Comment by kwoodson** [comment]
@JoelSpeed I have removed the webhooks code as per our discussion. This is still awaiting github.com/openshift/api/pull/1045 in order to keep in line with the api refactor.

**Comment by kwoodson** [comment]
@DahuK @menglingwei Please review the credentials request and verify that the permissions. If we can narrow them down, great. If this is sufficient then we can leave it as is.

https://github.com/openshift/machine-api-operator/pull/926/files#diff-88dafc65725c017015e420841dbf7ba59aadcfc05420b9d38d92bf8becd31dfdR221-R238

File: pkg/operator/config.go
Comment (JoelSpeed) [reply pkg/operator/config.go:119]: If it's already supported within the machine provider spec for alibaba, have they implemented a termination handler? Ref https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/spot-instances.md#termination-handler-design

**Review by DahuK — APPROVED** [review]

File: pkg/operator/config.go
Comment (menglingwei) [reply pkg/operator/config.go:119]: I have implemented a termination hanlder. I will submit a PR later.

Comment (menglingwei) [reply pkg/operator/config.go:119]: @kwoodson @JoelSpeed Tomorrow I will submit a new PR to openshift/cluster-api-provider-alibaba

**Review by menglingwei — APPROVED** [review]

**Review by menglingwei — COMMENTED** [review]
@kwoodson Alibaba Credential Request must have full ECS permissions? Do we only need the permissions used in cluster-api-provider-alibaba ?  I have listed the ECS permissions used in cluster-apI-provider-Alibaba below。

          - ecs:DeleteInstances
          - ecs:DescribeImages
          - ecs:DescribeInstances
          - ecs:DescribeSecurityGroups
          - ecs:RunInstances
          - ecs:StopInstances
          - ecs:TagResources

File: install/0000_30_machine-api-operator_00_credentials-request.yaml
Comment (mtulio) [inline install/0000_30_machine-api-operator_00_credentials-request.yaml:233]: Updating suggestion from [this comment](https://github.com/openshift/machine-api-operator/pull/926#pullrequestreview-814487051)
```suggestion
          - ecs:DeleteInstances
          - ecs:DescribeImages
          - ecs:DescribeInstances
          - ecs:DescribeSecurityGroups
          - ecs:RunInstances
          - ecs:StopInstances
          - ecs:TagResources
```

Comment (kwoodson) [reply install/0000_30_machine-api-operator_00_credentials-request.yaml]: Updated.

File: go.mod
Comment (JoelSpeed) [inline go.mod:7]: Updating this can commonly break the dependencies, which I think this has done, you'll probably need to pin these, we did this recently on Azure https://github.com/openshift/machine-api-provider-azure/blob/17e231c4fbc6eea84e171aa44b99bd7bfc83e58a/go.mod#L120

**Review by JoelSpeed — COMMENTED** [review]
Apart from the dependency issues, this is good to go, please see if you can fix the deps and let's get this merged

File: go.mod
Comment (JoelSpeed) [inline go.mod:30]: All the K8s versions need to match else you'll have issues with deps, please update them all to v0.23.0

**Comment by JoelSpeed** [comment]
Note to self, the Alibaba image is not yet in payload (https://issues.redhat.com/browse/ART-3535) and must be added before this can merge

File: go.mod
Comment (kwoodson) [reply go.mod:30]: Will update. Thanks

**Comment by kwoodson** [comment]
Updating caused updates to `pkg/operator/sync.go` which updated from a `Handler` to a `ProbeHandler`. Updated go.mod to satisfy the latest dependencies.

**Comment by kwoodson** [comment]
This is failing the tests because of 
```
error: unable to create a release: operator "machine-api-operator" contained an invalid image-references file: no input image tag named "alibaba-machine-controllers"
```
@elmiko @JoelSpeed Does this image reference look correct? I know the build is missing from Joel's comment https://github.com/openshift/machine-api-operator/pull/926#issuecomment-990168626

**Review by elmiko — COMMENTED** [review]
this is generally looking good to me, aside from the image issue
/lgtm

**Comment by kwoodson** [comment]
Libvirt test is failing because of a missing secret.
```
/bin/bash: line 9: /tmp/secret/openshift_gcp_compute_zone: No such file or directory
```

**Comment by elmiko** [comment]
should be fine @kwoodson , the only tests which haven't passed are non-required. if @JoelSpeed is ok to approve this we should be able to merge tomorrow.

**Comment by JoelSpeed** [comment]
We still can't merge this, https://issues.redhat.com/browse/ART-3535 is still in progress, AFAIK, waiting on ART

**Comment by kwoodson** [comment]
A rebase was required to fix go.mod dependencies. I have updated and redeployed a cluster with the latest changes:
```
NAME                                 STATUS   ROLES    AGE    VERSION
test-px2zh-master-0                  Ready    master   126m   v1.22.1+6859754
test-px2zh-master-1                  Ready    master   124m   v1.22.1+6859754
test-px2zh-master-2                  Ready    master   126m   v1.22.1+6859754
test-px2zh-worker-us-east-1a-555jl   Ready    worker   115m   v1.22.1+6859754
test-px2zh-worker-us-east-1b-2wph4   Ready    worker   114m   v1.22.1+6859754
test-px2zh-worker-us-east-1b-zlqvw   Ready    worker   111m   v1.22.1+6859754
NAME                                 PHASE     TYPE            REGION	ZONE         AGE
test-px2zh-master-0                  Running   ecs.g6.xlarge   us-east-1   us-east-1b   127m
test-px2zh-master-1                  Running   ecs.g6.xlarge   us-east-1   us-east-1a   127m
test-px2zh-master-2                  Running   ecs.g6.xlarge   us-east-1   us-east-1b   127m
test-px2zh-worker-us-east-1a-555jl   Running   ecs.g6.large    us-east-1   us-east-1a   121m
test-px2zh-worker-us-east-1b-2wph4   Running   ecs.g6.large    us-east-1   us-east-1b   121m
test-px2zh-worker-us-east-1b-zlqvw   Running   ecs.g6.large    us-east-1   us-east-1b   121m
```

**Comment by kwoodson** [comment]
Looking into why the unit tests failed.

**Comment by kwoodson** [comment]
@JoelSpeed @elmiko Please TAL. I believe the failures (e2e-aws) are failing all CI jobs in openshift. This PR is not the cause. I reviewed the unit test failures but was unsure as to why they are failing.

**Pull request merged**
